Restrict repository indexing by glob match

guillep2k opened the pull request:
### EDIT: This PR mutated along the way; to get the current state of this PR, please jump to [this comment](https://github.com/go-gitea/gitea/pull/7767#issuecomment-528701152).

Add an option in `app.ini` to restrict repository indexing by file extension:
```
[Indexer]
...
REPO_INDEXER_EXTENSIONS = go, c, cpp, h, idl, php, .
REPO_EXTENSIONS_LIST_EXCLUDE = false
```
It's not very powerful, but can help when indexes become polluted with derived files (e.g. `lss`, `sym`, `map`, etc.) so the search results are more relevant.

**REPO_INDEXER_EXTENSIONS** (default: empty) is a comma separated list of file extensions to include/exclude from the index. Only simple strings are supported (e.g. can't exclude `,v` from CVS). A dot (`.`) matches files with no extension. An empty list means "include all files", disregarding the value of **REPO_EXTENSIONS_LIST_EXCLUDE** (i.e. the current behaviour).
**REPO_EXTENSIONS_LIST_EXCLUDE** (default: `false`) defines whether **REPO_INDEXER_EXTENSIONS** is a list of extensions to exclude rather than include.

I didn't update `app.ini.sample` because I thought it migth be confusing for the user, since the default value is empty, anyway.

This PR is _related_ to #4831.

lafriks commented:
Might it not be better to set add extension not to index then?

guillep2k commented:
I suppose it depends on your particular case. I manage my company repository and there are projects ranging from VB6, C, C++, VB.Net, PHP, to Java and embedded. I still can't decide which method is more relevant to us. Lss, map, and sym are examples of files that definitely get in my way, but I'll probably decide for a list of known source code extensions to include rather than the opposite. There might be a lot of other use cases I can't think of, however.
I think that the **REPO_EXTENSIONS_LIST_EXCLUDE** flag is definitely useful, but there could be a **REPO_EXTENSIONS_LIST_INCLUDE** flag in its stead. The reason I decided for the former is that I could explain it better in fewer words 😅.

lafriks commented:
Include list would be usually more hard to maintain and you will most probably don't even know if you are missing something that is not indexed. Include list would probably have very rare use case imho

guillep2k commented:
Would you rather have a **REPO_EXTENSIONS_LIST_INCLUDE** flag, then?

lafriks commented:
Yes that would be less confusing

guillep2k commented:
OK, syntax is now as follows:
```
[Indexer]
...
REPO_INDEXER_EXTENSIONS = go, *.c, .cpp, h, idl, php, .
REPO_EXTENSIONS_LIST_INCLUDE = false
```
**REPO_INDEXER_EXTENSIONS** (default: empty) is a comma separated list of file extensions to **exclude** from the index. An empty list means _do not exclude files_ (i.e. the current behaviour).
**REPO_EXTENSIONS_LIST_INCLUDE** (default: `false`) if true, **REPO_INDEXER_EXTENSIONS** becomes a list of extensions to **include** rather than **exclude**.

**REPO_INDEXER_EXTENSIONS** now accepts `*.txt`, `.txt` and `txt` as synonyms. Dot (`.`) matches files with no extension.

lafriks commented:
Probably you could use https://github.com/gobwas/glob for wildcard support

guillep2k commented:
I've just added the `gobwas/glob` library for wildcard support. However, I think that power comes at the cost of added complexity.

In order to profit from the power of the glob patterns I'm using the full file path for matching (i.e. `lib/drivers/uart.c` vs. `uart.c`). This means patterns are not as intuitive as they could be. For instance, to match all the text files you need to use `**.txt` because the pattern `*.txt` only matches the files in the top level dir. On the other hand, it's possible to filter entire sub-directories using expressions like `data/**, **/data/**`.

I could have ignored the directory part of the path in the match and concentrate only on the actual file name, but then the library would be underutilized. Adding some examples to the docs would be very useful. I didn't want to pollute the `config-cheat-sheet` too much, so I kept it simple there.

To better reflect the effect of the filter, I've changed the names of the settings to `REPO_INDEXER_PATTERNS` and `REPO_PATTERNS_INCLUDE`.

**Note:** the last commit won't pass the build test because I don't know how to proceed with the added vendor. I've pushed just the code for your consideration.

PS: Sorry about the commit mess.

lafriks commented on models/repo_indexer.go:
```diff
@@ -232,6 +232,19 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
+			return false
+		}
+	}
 	return entry.IsRegular() || entry.IsExecutable()
```
this check should probably go before file pattern check. Also `if setting.Indexer.FilePatterns != nil {` should be reversed for cleaner code, something like:

```go
func isIndexable(entry *git.TreeEntry) bool {
	if !entry.IsRegular() && !entry.IsExecutable() {
		return false
	}
	if setting.Indexer.FilePatterns == nil {
		return true
	}
	// Check pattern match
	var found bool
	...
```

guillep2k replied:
OK, I'll turn that the other way around. About the new vendor, ¿should I simply "make vendor" and add anything that's modified in a new commit?

guillep2k replied:
Done!

lafriks commented:
why not use two patterns (include and exclude), so that it would be more flexible?

guillep2k commented:
> 
> 
> why not use two patterns (include and exclude), so that it would be more flexible?

@lafriks No problem. Back when it was only filtering by simple patterns it was not useful, but now with a glob library it makes sense. I'll work on it.

guillep2k commented:
@lafriks What do you think should take precedence, include or exclude?

guillep2k commented:
I propose exclude should have precedence over include. Example scenario:
```
REPO_INDEXER_INCLUDE = /project, /resources
REPO_INDEXER_EXCLUDE = .gitignore
```

lafriks commented:
Yes I agree that exclude should be taken into account first

guillep2k commented:
Added unit and integration tests. Please notice that the `gobwas/glob` library has a particular syntax, so my previous example:
```
REPO_INDEXER_INCLUDE = /project, /resources
REPO_INDEXER_EXCLUDE = .gitignore
```
should have been in fact written like:
```
REPO_INDEXER_INCLUDE = /project/**, /resources/**
REPO_INDEXER_EXCLUDE = **.gitignore
```

typeless commented on go.mod:
```diff
@@ -47,6 +47,7 @@ require (
 	github.com/go-sql-driver/mysql v1.4.1
 	github.com/go-swagger/go-swagger v0.20.1
 	github.com/go-xorm/xorm v0.7.7-0.20190822154023-17592d96b35b
+	github.com/gobwas/glob v0.2.3
```
Why does it have to be updated?

typeless replied:
is it because it is used in the tests?

guillep2k commented:
As this PR mutated quite a bit along the way, here is a summary of its functionality.

## What this PR does

Adds an option in `app.ini` to restrict _repository indexing_ by glob pattern matching using the `gobwas/glob` library:
```
[Indexer]
...
REPO_INDEXER_INCLUDE = project/**, resources/**
REPO_INDEXER_EXCLUDE = **.gitignore
```

It can help when indexes become polluted with derived files (e.g. `lss`, `sym`, `map`, etc.) so the search results are more relevant. It can also help reduce the index size.

**REPO_INDEXER_INCLUDE** (default: empty) is a comma separated list of glob patterns to **include** in the index. An empty list means "include all files".
**REPO_INDEXER_EXCLUDE** (default: empty) is a comma separated list of glob patterns to **exclude** from the index. Files that match this list will not be indexed. `REPO_INDEXER_EXCLUDE` takes precedence over `REPO_INDEXER_INCLUDE`.

Glob patterns as implemented by `gobwas/glob` are powerful and very fast, albeit a little complex. They work as follows:

* To match all files with .txt extension, no matter what directory, use `**.txt`.
* To match all files with .txt extension at the root level of the repository, use `*.txt`.
* To match all files inside `resources/bin` and below, use `resources/bin/**`.
* To match all files immediately inside `resources/bin`, use `resources/bin/*`.
* To match all Makefile files, use `**Makefile`.
* Matching a directory has no effect; `resources/bin` will not include/exclude files inside that directory; `resources/bin/**` will.
* All files and patterns are normalized to lower case.

lafriks commented on modules/setting/indexer.go:
```diff
@@ -51,10 +58,32 @@ func newIndexerService() {
+			}
+		}
+	}
+	if len(extarr) == 0 {
```
imho this is not needed

lafriks commented on models/repo_indexer.go:
```diff
@@ -234,7 +234,24 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
+			return false
+		}
+	}
+	if setting.Indexer.IncludePatterns == nil {
```
This is also not needed

lafriks commented on models/repo_indexer.go:
```diff
@@ -234,7 +234,24 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
+			return true
+		}
+	}
+	return false
```
```suggestion
	return len(setting.Indexer.IncludePatterns) == 0
```

guillep2k replied:
Done * 3

lafriks commented:
these last small things but otherwise great job! 🎉

guillep2k commented:
Now gobwas is one of the vendors, so it makes more sense to use it here.

lafriks commented:
make lg-tm work
